### PR TITLE
Feature/snowfalke retriever test

### PR DESCRIPTION
**Pull request by zhangchi1**
**What this PR does / why we need it**:
Update JDBC retriever in order to fit the snowflake as a historical database.

**Which issue(s) this PR fixes**:
1. Added snowflake as an option for historical retriever
2. Updated test case for snowflake retriever

**Next step**
1. Update JDBC configs
2. Update engine templates for different databases, such as snowflake, PostgreSQL

Ignore the typo of **snowflake** in the branch name.

Fixes #

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java [hunk 1]
```diff
@@ -138,11 +155,31 @@ private String exportResultsToDisk(Connection conn, String resultTable, String s
     String exportTableSqlQuery = null;
     try {
       Statement statement = conn.createStatement();
-      exportTableSqlQuery =
-          String.format(
-              "COPY %s TO '%s' WITH (DELIMITER E'\t', FORMAT CSV, HEADER);",
-              resultTable, exportPath);
-      statement.executeUpdate(exportTableSqlQuery);
+      if (this.className == "net.snowflake.client.jdbc.SnowflakeDriver") {
```
Comment (jmelinav) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java:157]: All the conditionals will be removed and code will be in specific querytemplater in the next PR right?

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java [hunk 2]
```diff
@@ -117,14 +132,16 @@ public HistoricalRetrievalResult getHistoricalFeatures(
...
+    // 7. export the result feature as a csv file to staging location
     String fileUri = exportResultsToDisk(conn, resultTable, stagingLocation);
     List<String> fileUris = new ArrayList<>();
+    // TODO: always return a single csv file?
     fileUris.add(fileUri);
     return HistoricalRetrievalResult.success(
         retrievalId, fileUris, ServingAPIProto.DataFormat.DATA_FORMAT_AVRO);
   }
 
+  // TODO: next step apply snowflake config
```
Comment (jmelinav) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: Remove your comments if already addressed.

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java (hunk 1 above)
Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java:157]: that's correct

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java (hunk 2 above)
Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: will do

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java [hunk 3]
```diff
@@ -326,8 +377,8 @@ String runBatchQuery(
       ResultSet rs = statement.executeQuery(timestampLimitSqlQuery);
 
       while (rs.next()) {
-        Timestamp min_ts = rs.getTimestamp("min"); // Get minimum timestamp
-        Timestamp max_ts = rs.getTimestamp("max"); // Get maximum timestamp
+        Timestamp min_ts = rs.getTimestamp(2); // Get minimum timestamp
```
Comment (jmelinav) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: Should this be reverted?

Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: the current implementation is correct. In the templater, "SELECT max(event_timestamp) as max, min(event_timestamp) as min from %s;". So Max is the first column and min is the second column. But I can clean up the order in order to make it easy to read.

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java [hunk 4]
```diff
@@ -326,8 +375,8 @@ String runBatchQuery(
       ResultSet rs = statement.executeQuery(timestampLimitSqlQuery);
 
       while (rs.next()) {
-        Timestamp min_ts = rs.getTimestamp("min"); // Get minimum timestamp
-        Timestamp max_ts = rs.getTimestamp("max"); // Get maximum timestamp
+        Timestamp max_ts = rs.getTimestamp(1); // Get maximum timestamp
```
Comment (jmelinav) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: Is this a valid change?

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/QueryTemplater.java [hunk 5]
```diff
@@ -68,9 +73,33 @@ public static String createEntityTableRowCountQuery(
...
-    return String.format(
-        "CREATE TABLE \"%s\" AS (SELECT %s FROM %s WHERE 1 = 2); ALTER TABLE \"%s\" ADD COLUMN event_timestamp TIMESTAMP, ADD COLUMN row_number SERIAL;",
-        destinationTable, featureSetTableSelectJoiner, featureSetTableFromJoiner, destinationTable);
+    List<String> createEntityTableRowCountQueries = new ArrayList<>();
+    createEntityTableRowCountQueries.add(
+        String.format(
+            "CREATE TABLE %s AS (SELECT %s FROM %s WHERE 1 = 2);",
+            destinationTable, featureSetTableSelectJoiner, featureSetTableFromJoiner));
+    if (className == "net.snowflake.client.jdbc.SnowflakeDriver") {
+      // 1st create a sequence
```
Comment (jmelinav) [inline storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/QueryTemplater.java]: remove commented lines if you don't need them.

File: storage/connectors/jdbc/src/test/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetrieverTest.java [hunk 6]
```diff
@@ -93,6 +90,7 @@ public void shouldRetrieveFromPostgresql() {
     HistoricalRetrievalResult postgresHisRetrievalResult =
         postgresqlFeatureRetriever.getHistoricalFeatures(
             retrievalId, datasetSource, featureSetRequests);
+    System.out.println(postgresHisRetrievalResult);
```
Comment (jmelinav) [inline storage/connectors/jdbc/src/test/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetrieverTest.java]: remove sysouts.

File: storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java (hunk 4 above)
Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: Yes, `rs.getTimestamp("max")` doesn't work in Snowflake. So I decided to use the column index number(index starts at 1).

Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: Using column index number works in Snowflake and Postgresql

Comment (zhangchi1) [reply storage/connectors/jdbc/src/main/java/feast/storage/connectors/jdbc/retriever/JdbcHistoricalRetriever.java]: Fixed the case sensitive issue. capitalized MAX, MIN  columns in order to fit both snowflake and Postgres databases.

**Pull request merged**